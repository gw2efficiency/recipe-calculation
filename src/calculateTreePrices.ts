import { RecipeTreeWithCraftFlags, RecipeTreeWithPrices, RecipeTreeWithQuantity } from './types'
import { CURRENCY_DECISION_PRICES } from './static/currencyDecisionPrices'

// Update the tree prices
export function calculateTreePrices(
  tree: RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>,
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithCraftFlags
export function calculateTreePrices(
  tree: RecipeTreeWithQuantity,
  itemPrices: Record<string, number>,
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithPrices
export function calculateTreePrices(
  tree: RecipeTreeWithQuantity | RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>,
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithPrices | RecipeTreeWithCraftFlags {
  // Calculate the buy prices
  let buyPriceEach = itemPrices[tree.id] || false
  if (tree.type === 'Currency') {
    // Coin is 1c per used quantity, the other currencies do not have a TP price
    buyPriceEach = tree.id === 1 ? 1 : false
  }
  const buyPrice = buyPriceEach ? tree.usedQuantity * buyPriceEach : false
  let craftResultPrice = buyPrice

  // Calculate the base decision price (may be overwritten later with the craft price)
  let decisionPriceEach = buyPriceEach || undefined
  if (tree.type === 'Currency') {
    if (customCurrencyPrices && typeof customCurrencyPrices[tree.id] === 'number') {
      decisionPriceEach = customCurrencyPrices[tree.id]
    } else {
      decisionPriceEach = CURRENCY_DECISION_PRICES[tree.id]
    }
  }
  let decisionPrice = decisionPriceEach ? tree.usedQuantity * decisionPriceEach : false

  if (!tree.components) {
    return {
      ...tree,
      components: undefined,
      buyPriceEach,
      buyPrice,
      decisionPrice,
      craftResultPrice,
      craftDecisionPrice: decisionPrice,
    }
  }

  // Calculate the tree prices traversal for the sub-components
  const components = tree.components.map((component) =>
    calculateTreePrices(component, itemPrices, customCurrencyPrices)
  )

  // Calculate the craft price out of the best prices
  const craftDecisionPrice = components.map((c) => c.decisionPrice || 0).reduce((a, b) => a + b, 0)
  const craftPrice = components.map((c) => c.craftResultPrice || 0).reduce((a, b) => a + b, 0)

  // Update the decision price of this tree segment to the craft price,
  // used to determine the craft price of the higher-up recipe
  if (
    !('craft' in tree && tree.craft === false) &&
    (('craft' in tree && tree.craft === true) ||
      !decisionPrice ||
      craftDecisionPrice < decisionPrice)
  ) {
    decisionPrice = craftDecisionPrice
    craftResultPrice = craftPrice
  }

  craftResultPrice = craftResultPrice || craftPrice
  decisionPrice = decisionPrice || craftDecisionPrice

  return {
    ...tree,
    components,
    buyPriceEach,
    buyPrice,
    craftPrice,
    decisionPrice,
    craftResultPrice,
    craftDecisionPrice,
  }
}
