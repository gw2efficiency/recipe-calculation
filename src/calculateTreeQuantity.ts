import { RecipeTreeWithCraftFlags, RecipeTree, RecipeTreeWithQuantity } from './types'

export function calculateTreeQuantity(
  amount: number,
  tree: RecipeTreeWithCraftFlags,
  availableItems?: Record<string, number>
): RecipeTreeWithCraftFlags
export function calculateTreeQuantity(
  amount: number,
  tree: RecipeTree,
  availableItems?: Record<string, number>
): RecipeTreeWithQuantity
export function calculateTreeQuantity(
  amount: number,
  tree: RecipeTree | RecipeTreeWithCraftFlags,
  availableItems: Record<string, number> = {}
) {
  // Make sure that we don't modify the passed-in object
  // We still want to work with a reference in the actual calculation
  // since the availableItems are a shared state for all sub-recipes
  return calculateTreeQuantityInner(amount, tree, { ...availableItems })
}

// Go through a recipe tree and set 'totalQuantity' based on the
// wanted amount and the output of recipes and sub-recipes
function calculateTreeQuantityInner(
  amount: number,
  tree: RecipeTree | RecipeTreeWithCraftFlags,
  availableItems: Record<string, number>,
  ignoreAvailable = false,
  nesting = 0
): RecipeTreeWithCraftFlags | RecipeTreeWithQuantity {
  const output = tree.output || 1

  // Calculate the total quantity needed
  let treeQuantity = amount * tree.quantity

  // Round amount to nearest multiple of the tree output
  treeQuantity = Math.ceil(treeQuantity / output) * output
  const totalQuantity = Math.round(treeQuantity)

  // If the item is available and the higher tree is not
  // bought or already available get as many items of it as possible
  // (This ignores the root node, because we *always* want to craft all of these)
  let availableQuantity = 0
  if (nesting > 0 && tree.type !== 'Currency' && !ignoreAvailable && availableItems[tree.id]) {
    availableQuantity = Math.min(availableItems[tree.id], totalQuantity)
    availableItems[tree.id] -= availableQuantity
  }
  const usedQuantity = totalQuantity - availableQuantity

  if (!tree.components) {
    return { ...tree, components: undefined, output, totalQuantity, usedQuantity }
  }

  // Get the amount of components that need to be crafted
  // e.g. a recipe outputs 10 and we need 20 -> 2x components
  const componentAmount = Math.ceil(usedQuantity / output)

  // Ignore available items in components if the tree
  // doesn't get crafted or is completely available anyway
  ignoreAvailable =
    ('craft' in tree && tree.craft === false) || usedQuantity === 0 || ignoreAvailable

  // Adjust the quantity for all tree's subcomponents
  const components = tree.components.map((component) => {
    return calculateTreeQuantityInner(
      componentAmount,
      component,
      availableItems,
      ignoreAvailable,
      ++nesting
    )
  })

  return { ...tree, components, output, totalQuantity, usedQuantity }
}
