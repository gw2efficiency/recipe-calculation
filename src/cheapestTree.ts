import { calculateTreeQuantity } from './calculateTreeQuantity'
import { calculateTreePrices } from './calculateTreePrices'
import { calculateTreeCraftFlags } from './calculateTreeCraftFlags'
import { RecipeTree, RecipeTreeWithCraftFlags, RecipeTreeWithPrices } from './types'
import {
  NestedRecipe,
  BasicCurrencyComponent,
  BasicItemComponent,
} from '@gw2efficiency/recipe-nesting'

export function cheapestTree(
  amount: number,
  tree: NestedRecipe,
  itemPrices: Record<string, number>,
  availableItems: Record<string, number> = {},
  forceBuyItems: Array<number> = [],
  valueOwnItems = false,
  userEfficiencyTiers: Record<string, string> = {
    '102306': '0',
    '102205': '0',
    '103049': '0',
  },
  customCurrencyPrices: Record<string, number>
): RecipeTreeWithCraftFlags {
  tree = applyEfficiencyTiersToTree(tree, userEfficiencyTiers)
  if (valueOwnItems) {
    const treeWithQuantityWithoutAvailableItems = calculateTreeQuantity(
      amount,
      tree as RecipeTree,
      {}
    )
    const treeWithPriceWithoutAvailableItems = calculateTreePrices(
      treeWithQuantityWithoutAvailableItems,
      itemPrices,
      customCurrencyPrices
    )

    const cheaperToBuyItemIds = getCheaperToBuyItemIds(treeWithPriceWithoutAvailableItems)

    // calculateTreeQuantity already checks for craft flags, so we can set them here
    tree = disableCraftForItemIds(tree, cheaperToBuyItemIds) as NestedRecipe
  }

  // Adjust the tree total and used quantities
  const treeWithQuantity = calculateTreeQuantity(amount, tree as RecipeTree, availableItems)

  // Set the initial craft flags based on the subtree prices
  const treeWithPrices = calculateTreePrices(treeWithQuantity, itemPrices, customCurrencyPrices)
  let treeWithCraftFlags = calculateTreeCraftFlags(treeWithPrices, forceBuyItems)

  // Force the root to be crafted
  treeWithCraftFlags = { ...treeWithCraftFlags, craft: true }

  // After the "craft" flags are set, update the used materials
  // to only be used for things that actually get crafted
  const treeWithQuantityPostFlags = calculateTreeQuantity(
    amount,
    treeWithCraftFlags,
    availableItems
  )

  // Recalculate the correct tree price
  return calculateTreePrices(treeWithQuantityPostFlags, itemPrices, customCurrencyPrices)
}

function getCheaperToBuyItemIds(
  tree: RecipeTreeWithPrices,
  ids: Array<number> = []
): Array<number> {
  // If buying the item costs less than the possible profit made by selling its craft components (with TP tax), we should buy instead of crafting
  if (
    typeof tree.craftDecisionPrice === 'number' &&
    typeof tree.buyPrice === 'number' &&
    tree.buyPrice < tree.craftDecisionPrice * 0.85
  ) {
    if (!ids.includes(tree.id)) {
      ids.push(tree.id)
    }
  }

  if (tree.components && Array.isArray(tree.components)) {
    tree.components.forEach((component) => getCheaperToBuyItemIds(component, ids))
  }

  return ids
}

type NestedRecipeAndBasicComponentWithCraftFlag = (
  | NestedRecipe
  | BasicItemComponent
  | BasicCurrencyComponent
) & {
  craft?: boolean
}

function disableCraftForItemIds(
  tree: NestedRecipeAndBasicComponentWithCraftFlag,
  ids: Array<number>
) {
  if (ids.includes(tree.id)) {
    tree = { ...tree, craft: false }
  }

  if ('components' in tree && Array.isArray(tree.components)) {
    tree.components = tree.components.map((component) =>
      disableCraftForItemIds(component as NestedRecipeAndBasicComponentWithCraftFlag, ids)
    )
  }

  return tree
}

function applyEfficiencyTiersToTree(
  tree: Omit<NestedRecipe, 'id'> & { id: number | null }, // FIXME Not sure why this can be null
  userEfficiencyTiers: Record<string, string>
): NestedRecipe {
  const id = tree.id ? tree.id.toString() : ''

  if (
    ['102306', '102205', '103049'].includes(id) &&
    tree.merchant &&
    tree.merchant.name.includes('Homestead Refinement')
  ) {
    const efficiencyTier = Number(userEfficiencyTiers[id])

    if (efficiencyTier > 0) {
      const component = { ...tree.components[0] }

      // Each efficiency tier lowers input by 50%, if it drops below one then doubles output
      component.quantity = component.quantity / (efficiencyTier * 2)

      // Bug: Onions are discounted by 75% with first tier
      if (component.id === 12142) {
        component.quantity = efficiencyTier === 1 ? 1 : 0.5
      }

      // Bug: Potatoes are not discounted with first tier
      if (component.id === 12135) {
        component.quantity = efficiencyTier === 1 ? 8 : 4
      }

      let updatedTree = { ...tree, output: component.quantity < 1 ? tree.output * 2 : tree.output }

      // Bug: Iron ore output also halves with second tier
      if (component.id === 19699 && efficiencyTier === 2) {
        updatedTree.output = updatedTree.output / 2
      }

      component.quantity = component.quantity < 1 ? 1 : component.quantity
      updatedTree = { ...updatedTree, components: [component, ...tree.components.slice(1)] }

      tree = updatedTree
    }
  }

  if ('components' in tree && Array.isArray(tree.components)) {
    tree = {
      ...tree,
      components: tree.components.map((component) =>
        applyEfficiencyTiersToTree(component as NestedRecipe, userEfficiencyTiers)
      ),
    }
  }

  return tree as NestedRecipe
}
