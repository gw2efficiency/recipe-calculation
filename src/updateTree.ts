import { calculateTreeQuantity } from './calculateTreeQuantity'
import { calculateTreePrices } from './calculateTreePrices'
import { RecipeTreeWithCraftFlags } from './types'

export function updateTree(
  amount: number,
  tree: RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>,
  availableItems: Record<string, number> = {},
  customCurrencyPrices: Record<string, number> = {}
) {
  // Update the tree total and used quantities
  const treeWithQuantity = calculateTreeQuantity(amount, tree, availableItems)

  // Recalculate the correct tree price
  return calculateTreePrices(treeWithQuantity, itemPrices, customCurrencyPrices)
}
