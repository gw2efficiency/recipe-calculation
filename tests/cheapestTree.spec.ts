import { NestedRecipe } from '@gw2efficiency/recipe-nesting'
import { cheapestTree } from '../src'
import { RecipeTreeWithCraftFlags } from '../src/types'

type SimplifiedTree = {
  id: number
  craft: boolean
  components?: Array<SimplifiedTree>
}

function simplifyTree(tree: RecipeTreeWithCraftFlags): SimplifiedTree {
  return {
    id: tree.id,
    craft: tree.craft,
    components: tree.components?.map((x) => simplifyTree(x)),
  }
}

describe('cheapestTree', () => {
  it('can calculate the cheapest tree correctly', () => {
    const recipeTree: NestedRecipe = {
      id: 1,
      type: 'Recipe',
      prerequisites: [{ type: 'Recipe', id: 124 }],
      min_rating: 300,
      disciplines: ['Amoursmith'],
      quantity: 1,
      output: 1,
      multipleRecipeCount: 1,
      components: [
        { id: 2, type: 'Item', quantity: 1 },
        {
          id: 3,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 5,
          multipleRecipeCount: 1,
          components: [
            { id: 4, type: 'Item', quantity: 2 },
            { id: 4, type: 'Currency', quantity: 200 },
            { id: 1, type: 'Currency', quantity: 100 },
          ],
        },
        {
          id: 5,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 5,
          components: [{ id: 6, type: 'Item', quantity: 2 }],
          multipleRecipeCount: 1,
        },
      ],
    }
    const prices = { 1: 10, 2: 42, 3: 10, 4: 10, 5: 1000, 6: 10 }

    const calculatedTree = cheapestTree(2, recipeTree, prices)
    expect(calculatedTree).toMatchSnapshot()
  })

  it('can calculate the cheapest tree correctly with available items', () => {
    const recipeTree: NestedRecipe = {
      id: 1,
      type: 'Recipe',
      prerequisites: [{ type: 'Recipe', id: 124 }],
      min_rating: 300,
      disciplines: ['Amoursmith'],
      quantity: 1,
      output: 1,
      multipleRecipeCount: 1,
      components: [
        {
          id: 7,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 3,
          components: [{ id: 4, type: 'Item', quantity: 1 }],
          multipleRecipeCount: 1,
        },
        {
          id: 3,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 5,
          components: [{ id: 4, type: 'Item', quantity: 2 }],
          multipleRecipeCount: 1,
        },
        {
          id: 5,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 2,
          components: [{ id: 6, type: 'Item', quantity: 100 }],
          multipleRecipeCount: 1,
        },
      ],
    }
    const prices = { 1: 10, 3: 100, 4: 10, 5: 25, 6: 1, 7: 1 }
    const availableItems = { 1: 100, 3: 7, 4: 3, 5: 4 }

    const calculatedTree = cheapestTree(2, recipeTree, prices, availableItems, [])
    expect(calculatedTree).toMatchSnapshot()
  })

  it('can calculate the cheapest tree correctly with force buy items', () => {
    const recipeTree: NestedRecipe = {
      id: 1,
      type: 'Recipe',
      prerequisites: [{ type: 'Recipe', id: 124 }],
      min_rating: 300,
      disciplines: ['Amoursmith'],
      quantity: 1,
      output: 1,
      multipleRecipeCount: 1,
      components: [
        {
          id: 3,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 5,
          components: [{ id: 4, type: 'Item', quantity: 2 }],
          multipleRecipeCount: 1,
        },
      ],
    }
    const prices = { 1: 10, 3: 100, 4: 10 }

    const calculatedTree = cheapestTree(2, recipeTree, prices, {}, [3])
    expect(calculatedTree).toMatchSnapshot()
  })

  it('can calculate the cheapest tree correctly with value own items', () => {
    const recipeTree: NestedRecipe = {
      id: 1,
      type: 'Recipe',
      prerequisites: [{ type: 'Recipe', id: 124 }],
      min_rating: 300,
      disciplines: ['Amoursmith'],
      quantity: 1,
      output: 1,
      multipleRecipeCount: 1,
      components: [
        {
          id: 3,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          quantity: 5,
          components: [{ id: 4, type: 'Item', quantity: 2 }],
          multipleRecipeCount: 1,
        },
      ],
    }

    const prices = { 1: 10, 3: 100, 4: 500 }
    const availableItems = { 3: 4, 4: 100 }

    const calculatedTreeNoOwn = cheapestTree(1, recipeTree, prices)
    const simpleTreeNoOwn = simplifyTree(calculatedTreeNoOwn)

    // Should make the same decisions as if we had no own materials
    const calculatedTree = cheapestTree(1, recipeTree, prices, availableItems, [], true)
    const simpleTree = simplifyTree(calculatedTree)
    expect(calculatedTree).toMatchSnapshot()
    expect(simpleTree).toEqual(simpleTreeNoOwn)

    // Should value own materials as "free" and make other decisions
    const calculatedTreeNoValue = cheapestTree(1, recipeTree, prices, availableItems, [], false)
    const simpleTreeNoValue = simplifyTree(calculatedTreeNoValue)
    expect(simpleTree).not.toEqual(simpleTreeNoValue)
  })

  it('can calculate the cheapest tree correctly with homestead efficiency tiers', () => {
    const recipeTree: NestedRecipe = {
      id: 1,
      type: 'Recipe',
      prerequisites: [{ type: 'Recipe', id: 124 }],
      min_rating: 300,
      disciplines: ['Amoursmith'],
      quantity: 5,
      output: 1,
      multipleRecipeCount: 1,
      components: [
        {
          id: 102205,
          type: 'Recipe',
          prerequisites: [{ type: 'Recipe', id: 123 }],
          output: 1,
          min_rating: 200,
          disciplines: ['Amoursmith'],
          merchant: {
            locations: [],
            name: '"Homestead Refinementâ€”Metal Forge"',
          },
          quantity: 25,
          components: [{ id: 4, type: 'Item', quantity: 4 }],
          multipleRecipeCount: 1,
        },
      ],
    }
    const prices = { 1: 10, 102205: 100, 4: 10 }

    const calculatedTree = cheapestTree(1, recipeTree, prices, {}, [], false, {
      '102306': '0',
      '102205': '2',
      '103049': '0',
    })
    expect(calculatedTree).toMatchSnapshot()
  })
})
