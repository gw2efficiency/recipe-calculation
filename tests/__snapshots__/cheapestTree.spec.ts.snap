// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cheapestTree can calculate the cheapest tree correctly 1`] = `
Object {
  "buyPrice": 20,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 84,
      "buyPriceEach": 42,
      "components": undefined,
      "craft": false,
      "craftDecisionPrice": 84,
      "craftResultPrice": 84,
      "decisionPrice": 84,
      "id": 2,
      "output": 1,
      "quantity": 1,
      "totalQuantity": 2,
      "type": "Item",
      "usedQuantity": 2,
    },
    Object {
      "buyPrice": 100,
      "buyPriceEach": 10,
      "components": Array [
        Object {
          "buyPrice": 200,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 200,
          "craftResultPrice": 200,
          "decisionPrice": 200,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 20,
          "type": "Item",
          "usedQuantity": 20,
        },
        Object {
          "buyPrice": false,
          "buyPriceEach": false,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 6000000,
          "craftResultPrice": false,
          "decisionPrice": 6000000,
          "id": 4,
          "output": 1,
          "quantity": 200,
          "totalQuantity": 2000,
          "type": "Currency",
          "usedQuantity": 2000,
        },
        Object {
          "buyPrice": 1000,
          "buyPriceEach": 1,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 1000,
          "craftResultPrice": 1000,
          "decisionPrice": 1000,
          "id": 1,
          "output": 1,
          "quantity": 100,
          "totalQuantity": 1000,
          "type": "Currency",
          "usedQuantity": 1000,
        },
      ],
      "craft": false,
      "craftDecisionPrice": 6001200,
      "craftPrice": 1200,
      "craftResultPrice": 100,
      "daily_purchase_cap": 0,
      "decisionPrice": 100,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 3,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 5,
      "totalQuantity": 10,
      "type": "Recipe",
      "usedQuantity": 10,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 10000,
      "buyPriceEach": 1000,
      "components": Array [
        Object {
          "buyPrice": 200,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 200,
          "craftResultPrice": 200,
          "decisionPrice": 200,
          "id": 6,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 20,
          "type": "Item",
          "usedQuantity": 20,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 200,
      "craftPrice": 200,
      "craftResultPrice": 200,
      "daily_purchase_cap": 0,
      "decisionPrice": 200,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 5,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 5,
      "totalQuantity": 10,
      "type": "Recipe",
      "usedQuantity": 10,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 384,
  "craftPrice": 384,
  "craftResultPrice": 384,
  "daily_purchase_cap": 0,
  "decisionPrice": 384,
  "disciplines": Array [
    "Amoursmith",
  ],
  "id": 1,
  "min_rating": 300,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [
    Object {
      "id": 124,
      "type": "Recipe",
    },
  ],
  "quantity": 1,
  "totalQuantity": 2,
  "type": "Recipe",
  "usedQuantity": 2,
  "weekly_purchase_cap": 0,
}
`;

exports[`cheapestTree can calculate the cheapest tree correctly with available items 1`] = `
Object {
  "buyPrice": 20,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 6,
      "buyPriceEach": 1,
      "components": Array [
        Object {
          "buyPrice": 60,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 60,
          "craftResultPrice": 60,
          "decisionPrice": 60,
          "id": 4,
          "output": 1,
          "quantity": 1,
          "totalQuantity": 6,
          "type": "Item",
          "usedQuantity": 6,
        },
      ],
      "craft": false,
      "craftDecisionPrice": 60,
      "craftPrice": 60,
      "craftResultPrice": 6,
      "daily_purchase_cap": 0,
      "decisionPrice": 6,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 7,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 3,
      "totalQuantity": 6,
      "type": "Recipe",
      "usedQuantity": 6,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 300,
      "buyPriceEach": 100,
      "components": Array [
        Object {
          "buyPrice": 30,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 30,
          "craftResultPrice": 30,
          "decisionPrice": 30,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 6,
          "type": "Item",
          "usedQuantity": 3,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 30,
      "craftPrice": 30,
      "craftResultPrice": 30,
      "daily_purchase_cap": 0,
      "decisionPrice": 30,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 3,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 5,
      "totalQuantity": 10,
      "type": "Recipe",
      "usedQuantity": 3,
      "weekly_purchase_cap": 0,
    },
    Object {
      "buyPrice": 0,
      "buyPriceEach": 25,
      "components": Array [
        Object {
          "buyPrice": 0,
          "buyPriceEach": 1,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 0,
          "craftResultPrice": 0,
          "decisionPrice": 0,
          "id": 6,
          "output": 1,
          "quantity": 100,
          "totalQuantity": 0,
          "type": "Item",
          "usedQuantity": 0,
        },
      ],
      "craft": false,
      "craftDecisionPrice": 0,
      "craftPrice": 0,
      "craftResultPrice": 0,
      "daily_purchase_cap": 0,
      "decisionPrice": 0,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 5,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 2,
      "totalQuantity": 4,
      "type": "Recipe",
      "usedQuantity": 0,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 36,
  "craftPrice": 36,
  "craftResultPrice": 36,
  "daily_purchase_cap": 0,
  "decisionPrice": 36,
  "disciplines": Array [
    "Amoursmith",
  ],
  "id": 1,
  "min_rating": 300,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [
    Object {
      "id": 124,
      "type": "Recipe",
    },
  ],
  "quantity": 1,
  "totalQuantity": 2,
  "type": "Recipe",
  "usedQuantity": 2,
  "weekly_purchase_cap": 0,
}
`;

exports[`cheapestTree can calculate the cheapest tree correctly with force buy items 1`] = `
Object {
  "buyPrice": 20,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 1000,
      "buyPriceEach": 100,
      "components": Array [
        Object {
          "buyPrice": 200,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 200,
          "craftResultPrice": 200,
          "decisionPrice": 200,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 20,
          "type": "Item",
          "usedQuantity": 20,
        },
      ],
      "craft": false,
      "craftDecisionPrice": 200,
      "craftPrice": 200,
      "craftResultPrice": 1000,
      "daily_purchase_cap": 0,
      "decisionPrice": 1000,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 3,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 5,
      "totalQuantity": 10,
      "type": "Recipe",
      "usedQuantity": 10,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 1000,
  "craftPrice": 1000,
  "craftResultPrice": 1000,
  "daily_purchase_cap": 0,
  "decisionPrice": 1000,
  "disciplines": Array [
    "Amoursmith",
  ],
  "id": 1,
  "min_rating": 300,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [
    Object {
      "id": 124,
      "type": "Recipe",
    },
  ],
  "quantity": 1,
  "totalQuantity": 2,
  "type": "Recipe",
  "usedQuantity": 2,
  "weekly_purchase_cap": 0,
}
`;

exports[`cheapestTree can calculate the cheapest tree correctly with homestead efficiency tiers 1`] = `
Object {
  "buyPrice": 50,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 12500,
      "buyPriceEach": 100,
      "components": Array [
        Object {
          "buyPrice": 1250,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 1250,
          "craftResultPrice": 1250,
          "decisionPrice": 1250,
          "id": 4,
          "output": 1,
          "quantity": 1,
          "totalQuantity": 125,
          "type": "Item",
          "usedQuantity": 125,
        },
      ],
      "craft": true,
      "craftDecisionPrice": 1250,
      "craftPrice": 1250,
      "craftResultPrice": 1250,
      "daily_purchase_cap": 0,
      "decisionPrice": 1250,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 102205,
      "merchant": Object {
        "locations": Array [],
        "name": "\\"Homestead Refinement—Metal Forge\\"",
      },
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 25,
      "totalQuantity": 125,
      "type": "Recipe",
      "usedQuantity": 125,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 1250,
  "craftPrice": 1250,
  "craftResultPrice": 1250,
  "daily_purchase_cap": 0,
  "decisionPrice": 1250,
  "disciplines": Array [
    "Amoursmith",
  ],
  "id": 1,
  "min_rating": 300,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [
    Object {
      "id": 124,
      "type": "Recipe",
    },
  ],
  "quantity": 5,
  "totalQuantity": 5,
  "type": "Recipe",
  "usedQuantity": 5,
  "weekly_purchase_cap": 0,
}
`;

exports[`cheapestTree can calculate the cheapest tree correctly with value own items 1`] = `
Object {
  "buyPrice": 10,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 100,
      "buyPriceEach": 100,
      "components": Array [
        Object {
          "buyPrice": 1000,
          "buyPriceEach": 500,
          "components": undefined,
          "craft": false,
          "craftDecisionPrice": 1000,
          "craftResultPrice": 1000,
          "decisionPrice": 1000,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 2,
          "type": "Item",
          "usedQuantity": 2,
        },
      ],
      "craft": false,
      "craftDecisionPrice": 1000,
      "craftPrice": 1000,
      "craftResultPrice": 100,
      "daily_purchase_cap": 0,
      "decisionPrice": 100,
      "disciplines": Array [
        "Amoursmith",
      ],
      "id": 3,
      "min_rating": 200,
      "multipleRecipeCount": 1,
      "output": 1,
      "prerequisites": Array [
        Object {
          "id": 123,
          "type": "Recipe",
        },
      ],
      "quantity": 5,
      "totalQuantity": 5,
      "type": "Recipe",
      "usedQuantity": 1,
      "weekly_purchase_cap": 0,
    },
  ],
  "craft": true,
  "craftDecisionPrice": 100,
  "craftPrice": 100,
  "craftResultPrice": 100,
  "daily_purchase_cap": 0,
  "decisionPrice": 100,
  "disciplines": Array [
    "Amoursmith",
  ],
  "id": 1,
  "min_rating": 300,
  "multipleRecipeCount": 1,
  "output": 1,
  "prerequisites": Array [
    Object {
      "id": 124,
      "type": "Recipe",
    },
  ],
  "quantity": 1,
  "totalQuantity": 1,
  "type": "Recipe",
  "usedQuantity": 1,
  "weekly_purchase_cap": 0,
}
`;
